presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java

```diff
@@ -0,0 +1,229 @@
+            for (Symbol symbol : sourceSymbols) {
+                outputSymbolBuilder.add(allocateNewSymbolFrom(symbol));
+            }
+            return outputSymbolBuilder.build();
+        }
+
+        private Symbol allocateNewSymbolFrom(Symbol symbol)
+        {
+            return allocateNewSymbolFrom(symbol.getName(), symbolAllocator.getTypes().get(symbol));
+        }
+
+        private AggregationNode createAggregationNode(IntersectNode node, List<Symbol> outputSymbols, UnionNode sourceNode, ImmutableMap.Builder<Symbol, FunctionCall> aggregationsBuilder, ImmutableMap.Builder<Symbol, Expression> outputProjections)
+        {
+            ImmutableMap.Builder<Symbol, Signature> functionsBuilder = ImmutableMap.builder();
+            ImmutableList.Builder<Symbol> groupByKeys = ImmutableList.builder();
+
+            List<Symbol> intersectOutputSymbols = node.getOutputSymbols();
+            Signature countSignature = new Signature("count", AGGREGATE, "bigint", "boolean");
+
+            for (int i = intersectOutputSymbols.size(); i < outputSymbols.size(); i++) {
+                Symbol countSymbol = allocateNewSymbolFrom("count", BIGINT);
+                aggregationsBuilder.put(countSymbol, new FunctionCall(QualifiedName.of("count"), ImmutableList.<Expression>of(new QualifiedNameReference(outputSymbols.get(i).toQualifiedName()))));
+                functionsBuilder.put(countSymbol, countSignature);
+            }
+
+            for (Symbol nodeSymbol : intersectOutputSymbols) {
+                int fieldIndex = intersectOutputSymbols.indexOf(nodeSymbol);
+                Symbol outputSymbol = outputSymbols.get(fieldIndex);
+                groupByKeys.add(outputSymbol);
+                outputProjections.put(nodeSymbol, new QualifiedNameReference(outputSymbol.toQualifiedName()));
+            }
+
+            return new AggregationNode(idAllocator.getNextId(),
+                    sourceNode,
+                    groupByKeys.build(),
+                    aggregationsBuilder.build(),
+                    functionsBuilder.build(),
+                    ImmutableMap.<Symbol, Symbol>of(),
+                    Step.SINGLE,
+                    Optional.empty(),
+                    1.0,
+                    Optional.empty());
+        }
+
+        private Symbol allocateNewSymbolFrom(String name, Type type)
+        {
+            return symbolAllocator.newSymbol(name, type);
+        }
+
+        private ProjectNode createProjectNode(IntersectNode intersectNode, List<PlanNode> sources, int sourceIndex)
+        {
+            ImmutableMap.Builder<Symbol, Expression> projectionAssignments = ImmutableMap.builder();
+            Map<Symbol, QualifiedNameReference> sourceSymbolMap = intersectNode.sourceSymbolMap(sourceIndex);
+
+            // add existing intersect symbols to projection
+            for (Symbol intersectSymbol : sourceSymbolMap.keySet()) {
+                Symbol symbol = allocateNewSymbolFrom(intersectSymbol);
```

Comment:
maybe we can reuse, I just allocated new ones for any new nodes I create so that we have unique symbols (and no conflicts). Let me look into this a bit and check if we can reuse